### PR TITLE
[SPARK-49444][SQL] Modified UnivocityParser to throw runtime exceptions caused by ArrayIndexOutOfBounds with more user-oriented messages

### PR DESCRIPTION
**Pull request by vladanvasi-db**
### What changes were proposed in this pull request?

I propose to catch and rethrow runtime `ArrayIndexOutOfBounds` exceptions in the `UnivocityParser` class - `parse` method, but with more user-oriented messages. Instead of throwing exceptions in the original format, I propose to inform the users which csv record caused the error.

### Why are the changes needed?

Proper informing of users' errors improves user experience. Instead of throwing `ArrayIndexOutOfBounds` exception without clear reason why it happened, proposed changes throw `SparkRuntimeException` with the message that includes original csv line which caused the error.

### Does this PR introduce _any_ user-facing change?

This PR introduces a user-facing change which happens when `UnivocityParser` parses malformed csv line with from the input. More specifically, the change is reproduces in the test case within `UnivocityParserSuite` when user specifies `maxColumns` in parser options and parsed csv record has more columns. Instead of resulting in `ArrayIndexOutOfBounds` like mentioned in the HMR ticket, users now get `SparkRuntimeException` with message that contains the input line which caused the error.

### How was this patch tested?

This patch was tested in `UnivocityParserSuite`. Test named "Array index out of bounds when parsing CSV with more columns than expected" covers this patch. Additionally, test for bad records in `UnivocityParser`'s `PERMISSIVE` mode is added to confirm that `BadRecordException` is being thrown properly.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Review by MaxGekk — CHANGES_REQUESTED** [review]
Please, remove the `ES-1203248` tag, and open OSS JIRA w/ proper description.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala [hunk 1]
```diff
@@ -273,11 +273,11 @@ class UnivocityParserSuite extends SparkFunSuite with SQLHelper {
...
-        input: String = "1,a",
-        dataSchema: StructType = StructType.fromDDL("i INTEGER, s STRING"),
-        requiredSchema: StructType = StructType.fromDDL("i INTEGER"),
-        filters: Seq[Filter],
-        expected: Option[InternalRow]): Unit = {
+               input: String = "1,a",
+               dataSchema: StructType = StructType.fromDDL("i INTEGER, s STRING"),
+               requiredSchema: StructType = StructType.fromDDL("i INTEGER"),
+               filters: Seq[Filter],
+               expected: Option[InternalRow]): Unit = {
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala:280]: wrong indentation

**Review by MaxGekk — COMMENTED** [review]
also cc @HyukjinKwon

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala [hunk 2]
```diff
@@ -323,6 +323,41 @@ class UnivocityParserSuite extends SparkFunSuite with SQLHelper {
...
+    val requiredSchema: StructType = StructType.fromDDL("i INTEGER, a STRING")
+    val options = new CSVOptions(Map("maxColumns" -> "2"), false, "UTC")
+    val filters = Seq()
+    val parser = new UnivocityParser(dataSchema, requiredSchema, options, filters)
+    checkError(
+      exception = intercept[SparkRuntimeException] {
+        parser.parse(input)
+      },
+      errorClass = "MALFORMED_CSV_RECORD",
+      parameters = Map("badRecord" -> "1,string,3.14,5,7"))
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala:358]: The error condition name seems misleading to me. The CSV record is not malformed, but we actually hit to some limits of implementation.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala [hunk 3]
```diff
@@ -98,7 +98,7 @@ class UnivocityParserSuite extends SparkFunSuite with SQLHelper {
   test("Throws exception for empty string with non null type") {
     val options = new CSVOptions(Map.empty[String, String], false, "UTC")
     val parser = new UnivocityParser(StructType(Seq.empty), options)
-    val exception = intercept[RuntimeException]{
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala:101]: Can we remove unrelated changes?

**Review by MaxGekk — CHANGES_REQUESTED** [review]
@vladanvasi-db Could you add an end-to-end test using public API, please.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala [hunk 4]
```diff
@@ -323,6 +323,41 @@ class UnivocityParserSuite extends SparkFunSuite with SQLHelper {
...
+      requiredSchema: StructType = StructType.fromDDL("i INTEGER, s STRING"),
+      options: Map[String, String] = Map("mode" -> "PERMISSIVE")): BadRecordException = {
+      val csvOptions = new CSVOptions(options, false, "UTC")
+      val parser = new UnivocityParser(dataSchema, requiredSchema, csvOptions, Seq())
+      intercept[BadRecordException] {
+        parser.parse(input)
+      }
+    }
+
+    // Bad record exception caused by conversion error
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala:339]: shall we test the same thing in the next test case that do not use PERMISSIVE mode?

Comment (vladanvasi-db) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala:339]: The `PERMISSIVE` mode will "Fill the row in a best-effort manner", so if it throws the exception, also the `FAILFAST` mode will throw the exception as well, so we do not need to check that one additionally. On the other hand, `DROPMALFORMED` mode will just drop the row and not throw the exception, so it is not applicable here when checking `BadRecordException`.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
